chromium_src/components/sync_device_info/device_info_sync_bridge_unittest.cc

```diff
@@ -192,5 +192,71 @@ TEST_F(DeviceInfoSyncBridgeTest,
   EXPECT_EQ(0u, bridge()->GetAllDeviceInfo().size());
 }
 
+TEST_F(DeviceInfoSyncBridgeTest, BraveExpireOldEntriesUponStartup) {
+  InitializeAndMergeInitialData(SyncMode::kFull);
+  ASSERT_EQ(1u, bridge()->GetAllDeviceInfo().size());
+  ASSERT_EQ(1, change_count());
+  ASSERT_FALSE(ReadAllFromStore().empty());
+
+  const DeviceInfoSpecifics specifics_old =
+      CreateSpecifics(2, base::Time::Now() - base::Days(57));
+
+  auto error = bridge()->ApplySyncChanges(bridge()->CreateMetadataChangeList(),
+                                          EntityAddList({specifics_old}));
+
+  ASSERT_FALSE(error);
+  ASSERT_EQ(2u, bridge()->GetAllDeviceInfo().size());
+  ASSERT_EQ(2, change_count());
+
+  // Reloading from storage should not expire the old entity
+  RestartBridge();
+  EXPECT_EQ(2u, bridge()->GetAllDeviceInfo().size());
+  // Make sure this is well persisted to the DB store.
+  EXPECT_THAT(ReadAllFromStore(),
+              UnorderedElementsAre(
+                  Pair(local_device()->GetLocalDeviceInfo()->guid(), _),
+                  Pair(specifics_old.cache_guid(), _)));
+}
+
+TEST_F(DeviceInfoSyncBridgeTest, BraveResetsProgressMarkerOnce) {
+  const DeviceInfoSpecifics specifics = CreateLocalDeviceSpecifics();
+  ModelTypeState model_type_state = StateWithEncryption("ekn");
+  model_type_state.mutable_progress_marker()->set_token("ABC");
+  WriteToStoreWithMetadata({specifics}, model_type_state);
+
+  InitializeBridge();
+
+  int current_call_number = 0;
+
+  // Wait until the metadata is loaded.
+  base::RunLoop run_loop;
+  EXPECT_CALL(*processor(), IsTrackingMetadata)
+      .Times(2)
+      .WillRepeatedly(Return(true));
+  EXPECT_CALL(*processor(), ModelReadyToSync)
+      .Times(2)
+      .WillRepeatedly([&run_loop, &current_call_number](
+                          std::unique_ptr<MetadataBatch> batch) {
+        ++current_call_number;
+        if (current_call_number == 1) {
+          // When model is loaded for the first time and the progress token
+          // was set then the token should be reset
+          EXPECT_TRUE(batch->GetModelTypeState().has_progress_marker());
+          EXPECT_FALSE(
+              batch->GetModelTypeState().progress_marker().has_token());
+        } else {
+          ASSERT_EQ(current_call_number, 2);
```

Comment:
are you sure this block is tested? RunLoop seems to only wait for the first call and then there is no second run waiting for second callback.

Reply:
@darkdh I re-wrote the test to be more clear and  those two cases are split now.
https://github.com/brave/brave-core/pull/12005/commits/f47ba5a9e99043d8232a5a11eaf45b01f4f0cbc4